pulsar-common/src/main/java/org/apache/pulsar/common/util/AsyncTokenBucket.java

```diff
@@ -0,0 +1,615 @@
+            this.targetFillFactorAfterThrottling = targetFillFactorAfterThrottling;
+            this.tokens = (long) (rateFunction.getAsLong() * initialTokensFactor);
+            tokens(false);
+        }
+
+        @Override
+        protected long getRatePeriodNanos() {
+            return ratePeriodNanosFunction.getAsLong();
+        }
+
+        @Override
+        protected long getTargetAmountOfTokensAfterThrottling() {
+            return (long) (getRate() * targetFillFactorAfterThrottling);
+        }
+
+        @Override
+        public long getCapacity() {
+            return capacityFactor == 1.0d ? getRate() : (long) (getRate() * capacityFactor);
+        }
+
+        @Override
+        public long getRate() {
+            return rateFunction.getAsLong();
+        }
+    }
+
+    /**
+     * A builder class for creating instances of {@link DynamicRateAsyncTokenBucket}.
+     */
+    public static class DynamicRateAsyncTokenBucketBuilder
+            extends AsyncTokenBucketBuilder<DynamicRateAsyncTokenBucketBuilder> {
+        protected LongSupplier rateFunction;
+        protected double capacityFactor = 1.0d;
+        protected double initialFillFactor = 1.0d;
+        protected LongSupplier ratePeriodNanosFunction;
+        protected double targetFillFactorAfterThrottling = 0.01d;
+
+        protected DynamicRateAsyncTokenBucketBuilder() {
+        }
+
+        public DynamicRateAsyncTokenBucketBuilder rateFunction(LongSupplier rateFunction) {
+            this.rateFunction = rateFunction;
+            return this;
+        }
+
+        public DynamicRateAsyncTokenBucketBuilder ratePeriodNanosFunction(LongSupplier ratePeriodNanosFunction) {
+            this.ratePeriodNanosFunction = ratePeriodNanosFunction;
+            return this;
+        }
+
+        public DynamicRateAsyncTokenBucketBuilder capacityFactor(double capacityFactor) {
+            this.capacityFactor = capacityFactor;
+            return this;
+        }
+
+        public DynamicRateAsyncTokenBucketBuilder initialFillFactor(double initialFillFactor) {
+            this.initialFillFactor = initialFillFactor;
+            return this;
+        }
+
+        public DynamicRateAsyncTokenBucketBuilder targetFillFactorAfterThrottling(
+                double targetFillFactorAfterThrottling) {
+            this.targetFillFactorAfterThrottling = targetFillFactorAfterThrottling;
+            return this;
+        }
+
+        @Override
+        public AsyncTokenBucket build() {
+            return new DynamicRateAsyncTokenBucket(this.capacityFactor, this.rateFunction,
+                    this.clockSource,
+                    this.ratePeriodNanosFunction, this.resolutionNanos,
+                    this.initialFillFactor,
+                    targetFillFactorAfterThrottling);
+        }
+    }
+
+    /**
+     * Interface for a clock source that returns a monotonic time in nanoseconds with a required precision.
+     */
+    public interface MonotonicClockSource {
```

Comment:
+1. The reason to put everything in a single file was to make AsyncTokenBucket self contained at first. I'm thinking of moving all AsyncTokenBucket files to a separate "qos" package.

One detail about this MonotonicClockSource interface is designed for AsyncTokenBucket's use case, so I'll keep it package private for now so that the usages can be controlled whenever that happens.

Reply:
Resolved by moving to separate file and renaming to MonotonicSnapshotClock